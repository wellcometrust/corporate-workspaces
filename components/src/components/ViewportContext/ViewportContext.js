// @flow
import React, { createContext, useEffect, useState } from 'react';
import type { Node } from 'react';

import mediaQueries from '../../utils/media-queries';

const defaultState = {
  isMobile: true,
  setIsMobile: () => {}
};

type ContextProps = {|
  isMobile: boolean,
  setIsMobile: (bool: boolean) => void
|};

type ProviderProps = {|
  children: Node
|};

const ViewportContext = createContext<ContextProps>(defaultState);

export const ViewportContextProvider = ({ children }: ProviderProps) => {
  const [state, setState] = useState({
    isMobile: true,
    setIsMobile: bool =>
      setState(prevState => ({ ...prevState, isMobile: bool }))
  });

  /**
   * useEffect hook with empty array argument is equivalent of
   * componentDidMount lifecycle method
   * used here to provide client side functionality
   * for setting up media query list (MQL) listener
   */
  useEffect(() => {
    const mql = window.matchMedia(mediaQueries.mediumMax);
    const handler = event => {
      state.setIsMobile(event.matches);
    };

    // Set initial state
    state.setIsMobile(mql.matches);

    // Set up listener
    mql.addListener(handler);

    // Remove listener on cleanup
    return () => mql.removeListener(handler);

    // Empty array as final argument ensures effect is only run on mount and unmount
  }, []);

  return (
    <ViewportContext.Provider value={state}>
      {children}
    </ViewportContext.Provider>
  );
};

export default ViewportContext;
