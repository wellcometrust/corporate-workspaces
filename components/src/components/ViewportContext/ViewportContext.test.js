import React from 'react';
import { mount } from 'enzyme';

import '../../test/mocks/match-media.mock';
import ViewportContext, { ViewportContextProvider } from './ViewportContext';

describe('<ViewportContextuProvider />', () => {
  let output;

  beforeEach(() => {
    output = mount(
      <ViewportContextProvider>
        <ViewportContext.Consumer>
          {({ isMobile }) => (
            <div className="test-isMobile">{isMobile.toString()}</div>
          )}
        </ViewportContext.Consumer>
      </ViewportContextProvider>
    );
  });

  afterEach(() => {
    output.unmount();
  });

  it('renders the component', () => {
    expect(output).toMatchSnapshot();
  });

  it('has `isMobile` property set to true', () => {
    expect(output.find('.test-isMobile').text()).toBe('true');
  });
});
