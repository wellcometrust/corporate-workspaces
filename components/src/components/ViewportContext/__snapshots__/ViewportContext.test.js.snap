// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ViewportContextuProvider /> renders the component 1`] = `ReactWrapper {}`;
